Sources/_StringProcessing/RegexDSL/DynamicCaptures.swift

```diff
@@ -17,28 +17,36 @@ extension Regex where Match == (Substring, DynamicCaptures) {
   }
 }
 
-public enum DynamicCaptures: Equatable {
-  case substring(Substring)
-  indirect case tuple([DynamicCaptures])
-  indirect case optional(DynamicCaptures?)
-  indirect case array([DynamicCaptures])
-
-  public static var empty: Self {
-    .tuple([])
+public struct DynamicCapture: Hashable {
```

Comment:
- This is a user-facing type but it does not have any public API. How would a user access it?
- Since it now represents a single capture rather than a tree, should we name it `AnyCapture` instead?
- We can have properties `public var base: Any` and `public var range: Range<String.Index>`.

Reply:
1. Is there any reason for DynamicCaptures to be public other than for the initializer taking one?
2. SGTM
3. Err, are back to exposing this to users?

Reply:
> Is there any reason for DynamicCaptures to be public other than for the initializer taking one?

The user would need to access the captured range and substring.

>  It would be `AnySubstringCapture`

Sure.

Reply:
> The user would need to access the captured range and substring. How else would the user do it?

We want API to get the number of captures, and for a given capture to get the slice/range/etc. I was thinking this would be more like static captures, where the more common approach API is on `RegexMatch` rather than `Match`. That is, `enum DynamicCaptures {}` would suffice as just a marker.

Or, we can have a dynamic capture type which holds the API. I'm not too particular on one approach over the other, just wondering what the rationale was or the tradeoffs.